examples/hello-world/main.cpp

```diff
@@ -1,137 +1,157 @@
+#include <slang-com-ptr.h>
+
+#include "vulkan-api.h"
 #include "examples/example-base/example-base.h"
 
-using namespace gfx;
-using namespace Slang;
+using Slang::ComPtr;
 
-// For the purposes of a small example, we will define the vertex data for a
-// single triangle directly in the source file. It should be easy to extend
-// this example to load data from an external source, if desired.
-//
-struct Vertex
+struct HelloWorldExample
 {
-    float position[3];
-    float color[3];
+    // The Vulkan functions pointers result from loading the vulkan library.
+    VulkanAPI vkAPI;
+
+    // Vulkan objects used in this example.
+    VkQueue queue;
+    VkCommandPool commandPool = VK_NULL_HANDLE;
+
+    // Input and output buffers.
+    VkBuffer inOutBuffers[3] = {};
+    VkDeviceMemory bufferMemories[3] = {};
+
+    const size_t inputElementCount = 16;
+    const size_t bufferSize = sizeof(float) * inputElementCount;
+
+    // We use a staging buffer allocated on host-visible memory to
+    // upload/download data from GPU.
+    VkBuffer stagingBuffer = VK_NULL_HANDLE;
+    VkDeviceMemory stagingMemory = VK_NULL_HANDLE;
+
+    VkDescriptorSetLayout descriptorSetLayout = VK_NULL_HANDLE;
+    VkPipelineLayout pipelineLayout = VK_NULL_HANDLE;
+    VkPipeline pipeline = VK_NULL_HANDLE;
+
+    // Initializes the Vulkan instance and device.
+    int initVulkanInstanceAndDevice();
+
+    // This function contains the most interesting part of this example.
+    // It loads the `hello-world.slang` shader and compile it using the Slang API
+    // into a SPIRV module, then create a Vulkan pipeline from the compiled shader.
+    int createComputePipelineFromShader();
+
+    // Creates the input and output buffers.
+    int createInOutBuffers();
+
+    // Sets up descriptor set bindings and dispatches the compute task.
+    int dispatchCompute();
+
+    // Reads back and prints the result of the compute task.
+    int printComputeResults();
+
+    // Main logic of this example.
+    int run();
+
+    ~HelloWorldExample();
+
 };
 
-static const int kVertexCount = 3;
-static const Vertex kVertexData[kVertexCount] =
+int main()
 {
-    { { 0,  0, 0.5 }, { 1, 0, 0 } },
-    { { 0,  1, 0.5 }, { 0, 0, 1 } },
-    { { 1,  0, 0.5 }, { 0, 1, 0 } },
-};
+    HelloWorldExample example;
+    return example.run();
+}
 
-// The example application will be implemented as a `struct`, so that
-// we can scope the resources it allocates without using global variables.
-//
-struct HelloWorld : public WindowedAppBase
+/************************************************************/
+/* HelloWorldExample Implementation */
+/************************************************************/
+
+int HelloWorldExample::run()
 {
+    RETURN_ON_FAIL(initVulkanInstanceAndDevice());
+    RETURN_ON_FAIL(createComputePipelineFromShader());
+    RETURN_ON_FAIL(createInOutBuffers());
+    RETURN_ON_FAIL(dispatchCompute());
+    RETURN_ON_FAIL(printComputeResults());
```

Comment:
I like the simplicity here. This could probably have some comments explaining the overall goal/flow of the application, and highlighting that the only Slang-specific bits are in `createComputePipelineFromShader()` (I assume).